File: 2023/kotlin/day03b/day03b.kt
import java.io.File

data class Position (val row: Int, val col: Int)

fun usage() {
    println("usage: java -jar <jarfile> <infile>")
    System.exit(1)
}

fun buildNumbers(contents: String): HashMap<Position, String> {
    var numberLocs = HashMap<Position, String>()
    var scanningNumber = false
    var number = StringBuilder()
    var currentPos = Position(row = -1, col = -1)
    var row = 0
    for ( line in contents.lines() ) {
        if ( line.isEmpty() ) {
            break
        }
        for ( col in 0..(line.length - 1) ) {
            var ch = line[col]
            if ( scanningNumber ) {
                if ( ch.isDigit() ) {
                    number.append(ch)
                } else {
                    numberLocs[currentPos] = number.toString()
                    scanningNumber = false
                    number = StringBuilder()
                }
            } else {
                if ( ch.isDigit() ) {
                    number.append(ch)
                    currentPos = Position(row = row, col = col)
                    scanningNumber = true
                }
            }
        }
        if ( scanningNumber ) {
            numberLocs[currentPos] = number.toString()
            scanningNumber = false
            number = StringBuilder()
        }
        row++
    }
    return numberLocs
}

fun buildParts(contents: String): HashMap<Position, Char> {
    var partLocs = HashMap<Position, Char>()
    var row = 0
    for ( line in contents.lines() ) {
        if ( line.isEmpty() ) {
            break
        }
        for ( col in 0..(line.length - 1) ) {
            val ch = line[col]
            if ( ch == '*' ) {
                partLocs[Position(row = row, col = col)] = ch;
            }
        }
        row++
    }
    return partLocs
}

fun checkParts(numberLocs: HashMap<Position, String>, partLocs: HashMap<Position, Char>): Int {
    var result = 0
    val neighbors = listOf(
        Position(row = -1, col = -1), Position(row = -1, col = 0), Position(row = -1, col = 1),
        Position(row = 0, col = -1), Position(row = 0, col = 1),
        Position(row = 1, col = -1), Position(row = 1, col = 0), Position(row = 1, col = 1),
    )
    for ( partLoc in partLocs.keys ) {
        var adjacentCount = 0
        var prod = 1
        for ( numberLocEntry in numberLocs.entries ) {
            val numberLoc = numberLocEntry.key
            val number = numberLocEntry.value
            val numberRow = numberLoc.row
            val numberColFirst = numberLoc.col
            val numberColLast = numberLoc.col + number.length - 1
            var found = false
            for ( neighbor in neighbors ) {
                val adjacent = Position(row = (partLoc.row + neighbor.row), col = (partLoc.col + neighbor.col))
                for ( numberCol in numberColFirst..numberColLast ) {
                    if ( numberRow == adjacent.row && numberCol == adjacent.col ) {
                        adjacentCount++
                        prod *= number.toInt()
                        found = true
                        break
                    }
                }
                if ( found ) {
                    break
                }
            }
        }
        if ( adjacentCount == 2 ) {
            result += prod
        }
    }
    return result
}

fun process(contents: String): Int {
    var numberLocs = buildNumbers(contents)
    var partLocs = buildParts(contents)
    return checkParts(numberLocs, partLocs)
}

fun main(args: Array<String>) {
    if ( args.size < 1 ) {
        usage()
    }
    val filename = args[0]
    val contents = File(filename).readText()
    val result = process(contents)
    println("result = $result")
}

File: 2023/javascript/day03b/day03b.js
const fs = require('fs');

function usage() {
    console.log(`usage: ${process.argv[0]} ${process.argv[1]} <file>`);
    process.exit(1);
}

function buildNumbers(contents) {
    var lines = contents.trim().split('\n');
    var numberLocs = new Map();
    var scanningNumber = false;
    var row = 0;
    var number = [];
    var currentPos = { row: -1,  col: -1};
    for ( var line of lines ) {
        for ( var col = 0; col < line.length; col++ ) {
            var ch = line.charAt(col);
            if ( scanningNumber ) {
                if ( ch >= '0' && ch <= '9' ) {
                    number.push(ch);
                } else {
                    numberLocs.set(currentPos, number.join(''));
                    scanningNumber = false;
                    number = [];
                }
            } else {
                if ( ch >= '0' && ch <= '9' ) {
                    number.push(ch);
                    currentPos = { row: row, col: col };
                    scanningNumber = true;
                }
            }
        }
        if ( scanningNumber ) {
            numberLocs.set(currentPos, number.join(''));
            scanningNumber = false;
            number = [];
        }
        row++;
    }
    return numberLocs;
}

function buildParts(contents) {
    var lines = contents.trim().split('\n');
    var partLocs = new Map();
    var row = 0;
    for ( var line of lines ) {
        for ( var col = 0; col < line.length; col++ ) {
            var ch = line[col];
            if ( ch == '*' ) {
                partLocs.set({ row: row, col: col }, ch);
            }
        }
        row++;
    }
    return partLocs;
}

function checkParts(numberLocs, partLocs) {
    var result = 0;
    var neighbors = [ { row: -1, col: -1 }, { row: -1, col: 0 }, { row: -1, col: 1 },
                      { row: 0, col: -1 }, { row: 0, col: 1 },
                      { row: 1, col: -1 }, { row: 1, col: 0 }, { row: 1, col: 1 } ];
    for ( var partLocEntry of partLocs.entries() ) {
        var partLoc = partLocEntry[0];
        var adjacentCount = 0;
        var prod = 1;
        for ( var numberLocEntry of numberLocs.entries() ) {
            var numberLoc = numberLocEntry[0];
            var number = numberLocEntry[1];
            var numRow = numberLoc.row;
            var numColFirst = numberLoc.col;
            var numColLast = numberLoc.col + number.length;
            var found = false;
            for ( var neighbor of neighbors ) {
                var adjacent = { row: partLoc.row + neighbor.row, col: partLoc.col + neighbor.col };
                for ( var numCol = numColFirst; numCol < numColLast; numCol++ ) {
                    if ( numRow == adjacent.row && numCol == adjacent.col ) {
                        adjacentCount++;
                        prod *= parseInt(number);
                        found = true;
                        break;
                    }
                }
                if ( found ) {
                    break;
                }
            }
        }
        if ( adjacentCount == 2 ) {
            result += prod;
        }
    }
    return result;
}

function processInput(contents) {
    numberLocs = buildNumbers(contents);
    partLocs = buildParts(contents);
    return checkParts(numberLocs, partLocs);
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    let filename = process.argv[2];
    let contents = fs.readFileSync(filename, 'utf8');
    let result = processInput(contents);
    console.log(`result = ${result}`);
}

main()
